zilliqa/src/contracts/mod.rs

```diff
@@ -39,220 +36,119 @@ pub mod native_token {
-        let vendor = root
-            .parent()
-            .unwrap()
-            .parent()
-            .unwrap()
-            .parent()
+    fn compile_all() {
+        let root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        let input = CompilerInput {
+            language: "Solidity".to_owned(),
+            sources: Source::read_all(
+                [
+                    "gas_price.sol",
+                    "native_token.sol",
+                    "shard.sol",
+                    "shard_registry.sol",
+                ]
+                .map(|c| format!("src/contracts/{c}")),
+            )
+            .unwrap(),
+            settings: Settings {
+                remappings: vec![Remapping {
+                    context: None,
+                    name: "@openzeppelin".to_owned(),
+                    path: "../vendor/openzeppelin-contracts/".to_owned(),
+                }],
+                optimizer: Optimizer {
+                    enabled: Some(true),
+                    runs: Some(4294967295),
+                    details: None,
+                },
+                output_selection: OutputSelection::complete_output_selection(),
+                evm_version: Some(EvmVersion::Shanghai),
+                ..Default::default()
+            },
+        };
+
+        let solc = Solc::find_or_install_svm_version("0.8.23")
             .unwrap()
-            .join("vendor");
+            .with_base_path(&root)
+            .args(["--allow-paths", "../vendor"]);
 
-        let output = Command::new(solc_path)
-            .current_dir(root)
-            .arg(contract)
-            .arg("@openzeppelin/=openzeppelin-contracts/")
-            .arg("--base-path")
-            .arg(".")
-            .arg("--include-path")
-            .arg(vendor)
-            .arg("--combined-json")
-            .arg("abi,bin,bin-runtime")
-            .arg("--evm-version=paris")
-            .output()
-            .unwrap();
+        let output = solc.compile_exact(&input).unwrap();
+        let output_file = root.join("src").join("contracts").join("compiled.json");
 
-        eprintln!("{}", std::str::from_utf8(&output.stderr).unwrap());
-        let combined_json: Value = serde_json::from_slice(&output.stdout).unwrap();
-        let contract = combined_json["contracts"][&json_key].clone();
+        if std::env::var_os("ZQ_CONTRACT_TEST_BLESS").is_some() {
+            let file = File::create(output_file).unwrap();
+            serde_json::to_writer(file, &output).unwrap();
 
-        let bin = contract["bin-runtime"].as_str().unwrap();
-        let expected_code = hex::decode(bin).unwrap();
-        assert_eq!(code.len(), expected_code.len()); // fail fast on a mismatch rather than dumping
-                                                     // full blobs into stdout
-        assert_eq!(code, expected_code);
+            panic!("compiled contracts updated, run this test again with `ZQ_CONTRACT_TEST_BLESS` unset");
```

Comment:
I read the PR description, but I was still initally confused the tests failed on the first run - I might put something like 'this is not a test failure, please rerun with...'

Reply:
Thinking about it more, there's no reason we should force you to run once with the env var set and then again with it unset to see a green tick locally. I'll just print this message and let the test pass in this branch.